Require const for variables from Ember properties

alexlafroscia opened the pull request:
**Good:**

```js
const { isNone } = Ember;
const isNone = Ember.isNone;  // Allowed by this rule, disallowed by other included ones
```

**Bad:**

```js
let { isNone } = Ember;
let isNone = Ember.isNone;
```

Closes #8

alexlafroscia commented:
For some reason, `npm test` was failing for me locally, saying that the "bad" `member.js` file that I added was not failing, but should be.  However, running

```bash
jscs -c lib/jscsrc.json tests/fixtures/rules/require-const-for-ember-properties/bad/member.js
```

directly would show that the file _does_ fail.  Is there a different way to test the file from the command line?

alexlafroscia commented:
Weird, the errors that show up in CI aren't showing up running `npm test` locally :-/

poteto commented on lib/rules/require-const-for-ember-properties.js:
```diff
@@ -0,0 +1,46 @@
+
+      if (variableFromEmber && node.kind !== 'const') {
+        errors.add(
+          'Always use `const` when making variables from Ember properties',
```
Maybe "Always use `const` when destructuring from the Ember class"

alexlafroscia replied:
Yeah, I like the sound of that better.  I had avoided that originally because you might be using something like

```js
let isNone = Ember.isNone;
```

which I know is also discouraged by Suave, but if someone had turned that rule off, then the above isn't really destructuring.  We could throw a different error in each case, though.

brzpegasus commented on tests/fixtures/rules/require-const-for-ember-properties/good/member.js:
```diff
@@ -0,0 +1,3 @@
+// jscs:disable requireObjectDestructuring
+const isNone = Ember.isNone;
```
Since I merged the `disallowDirectPropertyAccess` rule, this test will now fail as `Ember.isNone` is not allowed. You'll need to disable the rule here as well:

```js
// jscs:disable disallowDirectPropertyAccess
```

brzpegasus replied:
@alexlafroscia Or, if you're busy, I can make the change too =).

brzpegasus commented:
@alexlafroscia Finally getting back to this. If you have some time, could you please rebase and address this [comment](https://github.com/DockYard/ember-suave/pull/108/files#r65830265)? I'll merge as soon as it's done! Thanks.

alexlafroscia commented:
@brzpegasus Yup, totally! I just started a job where I need to get some permissions for OSS work so as soon as that's all settled I'll finish this up.

alexlafroscia commented:
@brzpegasus Sorry it took me so long to finish this up.  I rebased this PR off of `master`, and fixed the failing test 👍

brzpegasus commented on lib/rules/require-const-for-ember-properties.js:
```diff
@@ -0,0 +1,46 @@
+      if (variableFromEmber && node.kind !== 'const') {
+        errors.add(
+          'Always use `const` when making variables from Ember properties',
+          node.loc.start
```
In JSCS 3.0, this becomes `node` instead of `node.loc.start`.

brzpegasus commented on tests/fixtures/rules/require-const-for-ember-properties/bad/destructuring.js:
```diff
@@ -0,0 +1,5 @@
+let { isNone } = Ember;
+
+// Var
+var { isNone } = Ember;
```
Could you actually split up these two statements into separate files (one for `let` and one for `var`)? That way, we can be sure that both fail.

brzpegasus commented:
@alexlafroscia I uncovered a couple more things while testing this and left more comments. Also, you'll probably need to rebase against `master` again so you can get the caching fix from #128 (makes it very difficult to properly test rule changes otherwise!).

alexlafroscia commented:
Thanks! I'll take care of this stuff tonight.

alexlafroscia commented:
@brzpegasus Alright, so I split up those tests, made that JSCS compatibility change you mentioned and rebased off the current `master`.  Tests passed locally, so this should be ready for you to look at again.
